core/src/test/scala/unit/kafka/server/LogDirFailureTest.scala

```diff
@@ -191,12 +204,27 @@ class LogDirFailureTest extends IntegrationTestHarness {
     TestUtils.pollUntilAtLeastNumRecords(consumer, 1)
 
     // There should be no remaining LogDirEventNotification znode
-    assertTrue(zkClient.getAllLogDirEventNotifications.isEmpty)
+    if (quorum == "zk") {
+      assertTrue(zkClient.getAllLogDirEventNotifications.isEmpty)
+    }
 
-    // The controller should have marked the replica on the original leader as offline
-    val controllerServer = servers.find(_.kafkaController.isActive).get
-    val offlineReplicas = controllerServer.kafkaController.controllerContext.replicasInState(topic, OfflineReplica)
-    assertTrue(offlineReplicas.contains(PartitionAndReplica(new TopicPartition(topic, 0), leaderServerId)))
+    if (quorum == "kraft") {
+      waitUntilTrue(() => {
+        brokers.exists(broker => {
+          val hasOfflineDir = broker.asInstanceOf[BrokerServer].logDirFailureChannel.hasOfflineLogDir(failedLogDir.toPath.toString)
+          hasOfflineDir && broker.asInstanceOf[BrokerServer]
+            .replicaManager
+            .metadataCache
+            .getClusterMetadata(broker.clusterId, broker.config.interBrokerListenerName)
+            .partition(new TopicPartition(topic, 0)).offlineReplicas().map(_.id()).contains(originalLeaderServerId)
```

Comment:
@OmniaGM @soarez as I noticed, a metadata update won't happen in its current form of the JBOD feature as directories in ReplicationControlManager: https://github.com/apache/kafka/blob/c5ee82cab447b094ad7491200afa319515d5f467/metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java#L392
never get updated and all dirs have a default UNASSIGNED dir set. I'm guessing that AssignDirsToReplicas are never sent by the brokers upon topic creation or whenever they're supposed to be sent? Could this be a bug?
I'm trying to come up with a more closer RCA and a possible fix but I'm pinging you as I'm a bit unfamiliar with this area yet and you may have more much context on this. You can see the issue by putting a conditional breakpoint (which only stops if there are offline dirs) in handleDirectoriesOffline: https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java#L1400

Reply:
Hi @viktorsomogyi 

`ReplicaManager.maybeUpdateTopicAssignment` sends `AssignReplicasToDirs` as necessary after a partition is created. PR #14982 fixed an integration test — `kafka.api.PlaintextAdminIntegrationTest.testAlterReplicaLogDirs(String).quorum=kraft` — which relies on `AssignReplicasToDirs` being sent. So I'd expect that part to be working.

It could also be be that the log directory failure somehow isn't being propagated (via BrokerHeartbeatRequest), or that the metadata cache in the broker isn't being updated. To troubleshoot, I suggest verifying the following via logging/breakpoints:

1. Check that the assignment is being sent in `AssignmentsManager.onAssignment`
2. Check that the controller is receiving the assignment in `ReplicationControlManager.handleAssignReplicasToDirs`
3. Check that the directory failure is being propagated by the broker in `BrokerLifecycleManager.propagateDirectoryFailure`
4. Check that the controller is processing the directory failure in `ReplicationControlManager.handleDirectoriesOffline`
5. If all the above is happening, then there must be some issue in the up catching with metadata by the broker.

Reply:
Hey @viktorsomogyi, at least one of the issues here was this -> #14998

Reply:
Thanks, your referenced PR has indeed fixed these issues.